doc/languages-frameworks/maven.md

```diff
@@ -0,0 +1,351 @@
+# pick a repository derivation, here we will use buildMaven
+let repository = callPackage ./build-maven-repository.nix { };
+in stdenv.mkDerivation rec {
+  pname = "maven-demo";
+  version = "1.0";
+
+  src = builtins.fetchTarball "https://github.com/fzakaria/nixos-maven-example/archive/main.tar.gz";
+  buildInputs = [ maven ];
+
+  buildPhase = ''
+    echo "Using repository ${repository}"
+    mvn --offline -Dmaven.repo.local=${repository} package;
+  '';
+
+  installPhase = ''
+    install -Dm644 target/${pname}-${version}.jar $out/share/java
+  '';
+}
+```
+
+> We place the library in _$out/share/java_ since JDK package has a _stdenv setup hook_ that adds any JARs in the _share/java_ directories of the build inputs to the CLASSPATH environment.
+
+```bash
+❯ tree $(nix-build --no-out-link build-jar.nix)
+/nix/store/7jw3xdfagkc2vw8wrsdv68qpsnrxgvky-maven-demo-1.0
+└── share
+    └── java
+        └── maven-demo-1.0.jar
+
+2 directories, 1 file
+```
+
+## Runnable JAR
+
+You are able to run `mvn compile` but how do you get a runnable artifact without having to use something like building an UberJar ?
+
+We will use the same repository we built above (either _double invocation_ or _buildMaven_) to setup a CLASSPATH for our JAR.
+
+### CLASSPATH
+
+We will read the Maven repository and flatten it to a single list. This list will then be concatenated with the _CLASSPATH_ separator to create the full classpath.
+
+We make sure to provide this classpath to the _makeWrapper_.
+
+> readTree may be soon availble in nixpkgs via https://github.com/NixOS/nixpkgs/pull/101096
+
+```nix
+{ pkgs ? import <nixpkgs> { }, stdenv ? pkgs.stdenv, lib ? pkgs.lib
+, maven ? pkgs.maven, callPackage ? pkgs.callPackage
+, makeWrapper ? pkgs.makeWrapper, jre ? pkgs.jre }:
+let
+  readTree = dir:
+    lib.flatten (lib.mapAttrsToList (name: type:
+      if (type == "directory") then
+        readTree (dir + "/${name}")
+      else
+        dir + "/${name}") (builtins.readDir dir));
+
+  repository = callPackage ./build-maven-repository.nix { };
+  jars = lib.traceVal
+    (builtins.filter (path: (lib.hasSuffix ".jar" (toString path)))
+      (readTree repository));
+  classpath = lib.concatStringsSep ":" jars;
+in stdenv.mkDerivation rec {
+  pname = "maven-demo";
+  version = "1.0";
+
+  src = builtins.fetchTarball
+    "https://github.com/fzakaria/nixos-maven-example/archive/main.tar.gz";
+  buildInputs = [ maven makeWrapper ];
+
+  buildPhase = ''
+    echo "Using repository ${repository}"
+    mvn --offline -Dmaven.repo.local=${repository} package;
+  '';
+
+  installPhase = ''
+    mkdir -p $out/bin
+    mkdir -p $out/share/java
+
+    install -Dm644 target/${pname}-${version}.jar $out/share/java
+    # create a wrapper that will automatically set the classpath
+    # this should be the paths from the dependency derivation
+    makeWrapper ${jre}/bin/java $out/bin/${pname} \
+          --add-flags "-classpath $out/share/java/${pname}-${version}.jar:${classpath}" \
```

Comment:
`classpath` requires the repo to be built before evaluation can continue, but there's no need.

This could be implemented with `find` like
```bash
classpath=$(find ${repository} -name "*.jar" -printf ':%h/%f');
echo ${classpath#:} # remove the starting colon if any
```

It's also nice to set `repository` as a derivation attribute as in
```
stdenv.mkDerivation {
  # ...
  inherit repository;
  installPhase = ''
    # ...
    classpath=$(find $repository -name "*.jar" -printf ':%h/%f');
}
```

It has the side effect of adding a `repository` attribute that's accessible by other expressions and the repl.

Reply:
Same applies to `${jre}`.

Reply:
I agree with you if this was a real nixpkgs derivation for a real package but this is intended to be a working snippet to read through. Keeping it not fully declarative and overrideable might be the best approach.

Not sure I see your point about the _repository_;  doing 
`classpath=$(find ${repository} -name "*.jar" -printf ':%h/%f');` will require the derivation to be in the /nix/store or will build it.

Reply:
Resolved! thanks for the snippet as my bash sucks.